### DISCUSSION TITLE
Unifying or improving on event hooks & trace?

### DISCUSSION THREAD

**Discussion by lovelydinosaur** [encode/httpx#3379] [General]
We've currently got an [`event_hooks` API](https://www.python-httpx.org/advanced/event-hooks/) mirroring requests, with callbacks for `"request"` and `"response"` events.

However we've now also got a [much more comprehensive `"trace"` API](https://www.python-httpx.org/advanced/extensions/#trace).

Here are some use-cases we might want to be able to support...

* Set `request.started`, `request.complete`, `response.started`, `response.complete` timestamps.
* Monitor the % complete progress of a multipart upload.
* Implement SSL fingerprinting, raising exceptions when the fingerprint fails.

Are our existing APIs sufficient to be able to meet & document these use cases?
Are there improvements that we could make to provide a really neat unified API here?

**Comment by karpetrosyan** [comment]
I don’t find the `event_hook` API very useful. 
I prefer to use either custom transports or traces, as it seems I can do everything I need with them. 

Regarding the use cases

- We can subscribe to specific events and track the timestamp for each, not only these but also for SSL handshake, request sending, and many other processes.
- I believe we can’t currently trace the request-sending process because there is only a [single trace for the entire upload process](https://github.com/encode/httpcore/blob/0bfcee4cd266775c4d9ce2f89fc1b6afa8afc356/httpcore/_async/http11.py#L87). But seems like it's easy to fix, we just need to add a new trace in httpcore.

However, I would like to have some context for the traces so that they can communicate with each other and store useful information that can be utilized in other traces or even within the same trace at a different stage. I have an example of such a case [here](https://github.com/encode/httpcore/pull/932).

Initially, I thought about the context between stages. For example, `started` can store information such as a timestamp, while `failed` or `completed` can use that timestamp for metrics. This really adds extra flexibility and seems useful to me. However, now I’m considering a global context for the request. Here’s how it can look:

```python
import httpx


def log(event_name, info, context):
    _, event_name, stage = event_name.split(".")
    if event_name == "start_tls":
        context["tls_start_timestamp"] = 0


client = httpx.Client()

context = {}
extensions = {"trace_context": context, "trace": log}

client.get("https://www.example.com/", extensions=extensions)

print(context["tls_start_timestamp"])
```
